object-api-multipart.go

```diff
@@ -0,0 +1,492 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/minio/minio/pkg/probe"
+	"github.com/minio/minio/pkg/safe"
+	"github.com/skyrings/skyring-common/tools/uuid"
+)
+
+const (
+	minioMetaVolume    = ".minio"
+	slashPathSeparator = "/"
+)
+
+// checkLeafDirectory - verifies if a given path is leaf directory if
+// yes returns all the files inside it.
+func (o objectAPI) checkLeafDirectory(prefixPath string) (isLeaf bool, fis []FileInfo) {
+	var allFileInfos []FileInfo
+	for {
+		fileInfos, eof, e := o.storage.ListFiles(minioMetaVolume, prefixPath, "", false, 1000)
+		if e != nil {
+			break
+		}
+		allFileInfos = append(allFileInfos, fileInfos...)
+		if eof {
+			break
+		}
+	}
+	for _, fileInfo := range allFileInfos {
+		if fileInfo.Mode.IsDir() {
+			isLeaf = false
+			return isLeaf, nil
+		}
+		fileName := path.Base(fileInfo.Name)
+		if !strings.Contains(fileName, ".") {
+			fis = append(fis, fileInfo)
+		}
+	}
+	isLeaf = true
+	return isLeaf, fis
+}
+
+// ListMultipartUploads - list multipart uploads.
+func (o objectAPI) ListMultipartUploads(bucket, prefix, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (ListMultipartsInfo, *probe.Error) {
+	result := ListMultipartsInfo{}
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return ListMultipartsInfo{}, probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	if !IsValidObjectPrefix(prefix) {
+		return ListMultipartsInfo{}, probe.NewError(ObjectNameInvalid{Bucket: bucket, Object: prefix})
+	}
+	// Verify if delimiter is anything other than '/', which we do not support.
+	if delimiter != "" && delimiter != slashPathSeparator {
+		return ListMultipartsInfo{}, probe.NewError(UnsupportedDelimiter{
+			Delimiter: delimiter,
+		})
+	}
+	// Verify if marker has prefix.
+	if keyMarker != "" && !strings.HasPrefix(keyMarker, prefix) {
+		return ListMultipartsInfo{}, probe.NewError(InvalidMarkerPrefixCombination{
+			Marker: keyMarker,
+			Prefix: prefix,
+		})
+	}
+	if uploadIDMarker != "" {
+		if strings.HasSuffix(keyMarker, slashPathSeparator) {
+			return result, probe.NewError(InvalidUploadIDKeyCombination{
+				UploadIDMarker: uploadIDMarker,
+				KeyMarker:      keyMarker,
+			})
+		}
+		id, e := uuid.Parse(uploadIDMarker)
+		if e != nil {
+			return result, probe.NewError(e)
+		}
+		if id.IsZero() {
+			return result, probe.NewError(MalformedUploadID{
+				UploadID: uploadIDMarker,
+			})
+		}
+	}
+
+	recursive := true
+	if delimiter == slashPathSeparator {
+		recursive = false
+	}
+	result.IsTruncated = true
+	newMaxUploads := 0
+	prefixPath := bucket + slashPathSeparator + prefix // do not use filepath.Join so that we retain trailing '/' if any
```

Comment:
is it valid to have `mybucket//prefix/`? I think its not a good idea to do string cat for paths.